drivers/flash/flash_ipuc/flash_ipuc.c

```diff
@@ -0,0 +1,458 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <string.h>
+
+#include <zephyr/drivers/flash.h>
+#include <drivers/flash/flash_ipuc.h>
+#include <suit_ipuc.h>
+#include <suit_plat_decode_util.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(flash_ipuc, CONFIG_FLASH_IPUC_LOG_LEVEL);
+
+#define FLASH_WRITE_BLOCK_SIZE DT_PROP(DT_CHOSEN(zephyr_flash), write_block_size)
+#define WRITE_BLOCK_SIZE       FLASH_WRITE_BLOCK_SIZE
+#define ERASE_VALUE	       0xff
+
+BUILD_ASSERT(WRITE_BLOCK_SIZE > 0, "zephyr,flash: write_block_size expected to be non-zero");
+
+#define IPUC_MEM_COMPONENT_MAX_SIZE 32
+
+struct ipuc_context {
+	struct zcbor_string component_id;
+	uint8_t component_id_buf[IPUC_MEM_COMPONENT_MAX_SIZE];
+	size_t address;
+	size_t size;
+	bool read_access;
+};
+
+#define DEFINE_NRF_IPUC_DATA(x, _)                                                                 \
+	static struct ipuc_context ipuc_context_data_##x = {                                       \
+		.component_id = {NULL, 0},                                                         \
+		.address = 0,                                                                      \
+		.size = 0,                                                                         \
+	}
+
+#define DEFINE_NRF_IPUC(x, _)                                                                      \
+	DEVICE_DEFINE(flash_nrf_ipuc_##x, "nrf_ipuc_" #x, NULL, /* init */                         \
+		      NULL,					/* pm */                           \
+		      &ipuc_context_data_##x, NULL,		/* config */                       \
+		      POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, &nrf_ipuc_api)
+
+#define DEFINE_NRF_IPUC_REF(x, _) &__device_flash_nrf_ipuc_##x
+
+static int nrf_ipuc_read(const struct device *dev, off_t offset, void *data, size_t len)
+{
+	struct ipuc_context *ctx = NULL;
+
+	if ((dev == NULL) || (data == NULL)) {
+		return -EINVAL;
+	}
+
+	ctx = (struct ipuc_context *)dev->data;
+	if (!ctx->read_access) {
+		return -EACCES;
+	}
+
+	LOG_DBG("read: %p:%zu", (void *)offset, len);
+
+	if (offset + len > ctx->size) {
+		return -ENOMEM;
+	}
+
+	memcpy(data, (void *)(ctx->address + offset), len);
```

Comment:
This does not seem to be the right thing to do...
If we attempt to read an address in the secure part in the non-secure context or we attempt to read radio core memory from the application core we will get some kind of a hard fault.

Rather we should protect ourselves against such scenarios by doing some checks and returning an error earlier, though I do not know right away how to do it...

By the way, why is reading from IPUC supported anyway? Checking of digests should be done using dedicated functions, couldn't this function implementation be simply returning an error?

Reply:
Reading is necessary to implement cache over IPUC driver.
The `ctx->read_access` flag should protect from unexpected hard faults.
There is no such thing like "secure" or "non-secure" SUIT manifest. I do not thing that the current design is (at all) able to handle the S/NS split in other way than including SUIT in the S part.